File: Exercism/c/clock/src/clock.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include "clock.h"

void clock(time_text_t time_text, int hour, int minute) {
    if (hour < 0)
        hour = 24+(hour%24);

    if (minute >= 60) {
        hour += minute/60;
        minute -= minute/60*60;
    }

    if (minute < 0) {
        hour -= ceil((minute*-1.0)/60.0);

        if (hour < 0)
            hour += 24;

        minute += floor(minute*-1.0/60.0)*60;
        minute = 60+minute;

        if (hour < 0)
            hour = 24+(hour%24);

    }

    sprintf(time_text, "%02d:%02d", hour%24, minute);
}

void clock_add(time_text_t time_text, int minute_offset) {
    char hour_s[2];
    strncpy(hour_s, time_text, 2);
    int hour = atoi(hour_s);

    char minute_s[2];
    strncpy(minute_s, time_text+3, 2);
    int minute = atoi(minute_s);

    hour += minute_offset/60;
    minute_offset -= minute_offset/60*60;
    minute += minute_offset;

    clock(time_text, hour, minute);
}

File: Exercism/ruby/clock/clock.rb
# frozen_string_literal: true

# Clock ...
class Clock
  def to_s
    format('%<hour>02d:%<minute>02d', hour: @hour, minute: @minute)
  end

  def -(other)
    Clock.new(hour: @hour-other.hour, minute: @minute-other.minute)
  end

  def +(other)
    Clock.new(hour: @hour+other.hour, minute: @minute+other.minute)
  end

  def ==(other)
    @hour == other.hour && @minute == other.minute
  end

  def initialize(hour: 0, minute: 0)
    hour = 24 + (hour % 24) if hour.negative?

    if minute >= 60
      hour += minute / 60
      minute -= minute / 60 * 60
    end

    if minute.negative?
      hour -= ((minute * -1.0) / 60.0).ceil
      hour += 24 if hour.negative?
      minute += (minute * -1.0 / 60.0).floor * 60
      minute += 60 unless minute.zero?
      hour = 24 + (hour % 24) if hour.negative?
    end

    @hour = hour % 24
    @minute = minute
  end

  def hour
    @hour
  end

  def minute
    @minute
  end
end
